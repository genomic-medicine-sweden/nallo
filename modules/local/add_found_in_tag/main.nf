process ADD_FOUND_IN_TAG {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/bcftools:1.20--h8b25389_0':
        'biocontainers/bcftools:1.20--h8b25389_0' }"

    input:
    tuple val(meta), path(vcf, stageAs: "?/*"), path(index, stageAs: "?/*")
    val(variant_caller)

    output:
    tuple val(meta), path("*.{vcf,vcf.gz,bcf,bcf.gz}"), emit: vcf
    tuple val(meta), path("*.tbi")                    , emit: tbi, optional: true
    tuple val(meta), path("*.csi")                    , emit: csi, optional: true
    path "versions.yml"                               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def extension = args2.contains("--output-type b") || args2.contains("-Ob") ? "bcf.gz" :
                    args2.contains("--output-type u") || args2.contains("-Ou") ? "bcf" :
                    args2.contains("--output-type z") || args2.contains("-Oz") ? "vcf.gz" :
                    args2.contains("--output-type v") || args2.contains("-Ov") ? "vcf" :
                    "vcf"

    """
    bcftools view \\
        $args \\
        --threads $task.cpus \\
        $vcf |\\
    awk '
        # Print all existing header lines
        /^##/ {
            print;
            next;
        }

        # Add a INFO header line before the #CHROM line, then print the #CHROM line
        /^#CHROM/ {
            printf("##INFO=<ID=FOUND_IN,Number=1,Type=String,Description=\\"Program that called the variant\\">\\n");
            print;
            next;
        }

        # Then for all other lines, append FOUND_IN=variant_caller to the INFO field (column 8)
        {
            OFS="\\t";
            \$8 = sprintf("%s;FOUND_IN=${variant_caller}", \$8 );
            print;
        }
    ' |\\
    bcftools view \\
        $args2 \\
        --threads $task.cpus \\
        --output ${prefix}.${extension}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
        gawk: \$(awk -Wversion | sed '1!d; s/.*Awk //; s/,.*//')
    END_VERSIONS
    """

    stub:
    def args2 = task.ext.args2 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def extension = args2.contains("--output-type b") || args2.contains("-Ob") ? "bcf.gz" :
                    args2.contains("--output-type u") || args2.contains("-Ou") ? "bcf" :
                    args2.contains("--output-type z") || args2.contains("-Oz") ? "vcf.gz" :
                    args2.contains("--output-type v") || args2.contains("-Ov") ? "vcf" :
                    "vcf"
    def index = args2.contains("--write-index=tbi") || args2.contains("-W=tbi") ? "tbi" :
                args2.contains("--write-index=csi") || args2.contains("-W=csi") ? "csi" :
                args2.contains("--write-index") || args2.contains("-W") ? "csi" :
                ""
    def create_cmd = extension.endsWith(".gz") ? "echo '' | gzip >" : "touch"
    def create_index = extension.endsWith(".gz") && index.matches("csi|tbi") ? "touch ${prefix}.${extension}.${index}" : ""

    """
    ${create_cmd} ${prefix}.${extension}
    ${create_index}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
        busybox-awk: \$(awk -Wversion | sed -n 's/.*v\\([^ ]*\\) (.*/\\1/p')
    END_VERSIONS
    """
}
