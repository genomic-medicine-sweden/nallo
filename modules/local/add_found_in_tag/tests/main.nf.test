nextflow_process {

    name "Test Process ADD_FOUND_IN_TAG"
    script "../main.nf"
    process "ADD_FOUND_IN_TAG"

    tag "modules"
    tag "modules_local"
    tag "bcftools"
    tag "add_found_in_tag"

    test("sarscov2 - [vcf, []], variant_caller") {

        when {
            process {
                """
                input[0] = [
                    [ id:'out', single_end:false ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz.tbi', checkIfExists: true)
                ]
                input[1] = 'variant_caller'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert file(process.out.vcf.get(0).get(1)).readLines().any { it.contains("FOUND_IN=variant_caller") } },
                { assert snapshot(
                    path(process.out.vcf.get(0).get(1)).vcf.variantsMD5,
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - [vcf, []], variant_caller -stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'out', single_end:false ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz.tbi', checkIfExists: true)
                ]
                input[1] = 'variant_caller'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
