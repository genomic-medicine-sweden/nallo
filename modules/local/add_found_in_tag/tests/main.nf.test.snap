{
    "sarscov2 - [vcf, []], variant_caller": {
        "content": [
            "7ce88babfa8801b2a913faf40c6f7edd",
            [
                "versions.yml:md5,366aba7250a7ed36b3a0fa4b534ad33d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-09-17T12:12:10.307079371"
    },
    "sarscov2 - [vcf, []], variant_caller -stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": [
                                [
                                    [
                                        "HG002_Revio_phased.bam:md5,64e1d1ed3d180589193c13f2c94d1098",
                                        "HG002_Revio_phased.bam.bai:md5,bb768f48be5144262c0b713ef183d152"
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "HG002_Revio.asm.bp.hap1.p_ctg.assembly_summary:md5,e4da3107f2421d42c6e2f59bff4e47d9",
                                            "HG002_Revio.asm.bp.hap1.p_ctg.fasta.gz:md5,eff16bf92abff02c36cceb2eb5a9871c",
                                            "HG002_Revio.asm.bp.hap2.p_ctg.assembly_summary:md5,bf908610472e602fd986b9d4a4d58b9e",
                                            "HG002_Revio.asm.bp.hap2.p_ctg.fasta.gz:md5,e28c886baffe24c1a2972a1d95b1b250"
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "HG002_Revio.dip.bed:md5,4e3cde029bceaa9363402e997f12f280",
                                            "HG002_Revio.dip.vcf.gz:md5,9259613d7b01347b2d817d4e547c4203",
                                            "HG002_Revio.hap1.bam:md5,5c739e8b3c2f659c3847498848280bbf",
                                            "HG002_Revio.hap1.bam.bai:md5,a57472fa9365470db09f5ea8c7a7420a",
                                            "HG002_Revio.hap1.bed:md5,3d75c9538e748a7672b934bdbde90df3",
                                            "HG002_Revio.hap1.paf.gz:md5,46a3056ecd833d9a932c53fc46a3cafc",
                                            "HG002_Revio.hap1.sam.gz:md5,0a161b44c29c0581ada87fee50be38c2",
                                            "HG002_Revio.hap1.var.gz:md5,173218f6c3f5f067e0932a8331f12e3d",
                                            "HG002_Revio.hap2.bam:md5,df63e7bb5dcee6ca5e731272dad0eb99",
                                            "HG002_Revio.hap2.bam.bai:md5,ed8184223edea0297c7f9283ea7927bf",
                                            "HG002_Revio.hap2.bed:md5,9ac411d49f08a8ec70e4792890e9faf9",
                                            "HG002_Revio.hap2.paf.gz:md5,a99e451e4af96a7ab120cedd49eea618",
                                            "HG002_Revio.hap2.sam.gz:md5,410acad899eda2f0a4a3e1eb4acf81cb",
                                            "HG002_Revio.hap2.var.gz:md5,53dc16fc893a529a95067da161b4dad6",
                                            "HG002_Revio.pair.vcf.gz:md5,a390558ab50ce0dc9efec070b1c70c22"
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "HG002_Revio.HG002_Revio.copynum.bedgraph:md5,517bc59c1b235490c79aa8319437b033",
                                            "HG002_Revio.HG002_Revio.depth.bw:md5,20757c985f4713b8568dee05765db701",
                                            "HG002_Revio.HG002_Revio.maf.bw:md5,4699d7a664277863f06eb48c3cba3c41",
                                            "HG002_Revio.HG002_Revio.vcf.gz:md5,771f7d9ecbd161dedfbdd053ceff0d52",
                                            "HG002_Revio.HG002_Revio.vcf.gz.tbi:md5,65d6930f79e626a5542091b7a419dc2b",
                                            "HG002_Revio.log:md5,3b47302d7e8e0ffac4841123de74ffb3"
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            [
                                                
                                            ],
                                            [
                                                
                                            ]
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        "HG002_Revio.paraphase.bam:md5,839cab6fb9b453b36975e9ab650d2552",
                                        "HG002_Revio.paraphase.bam.bai:md5,7146463306700bc589b4085c3404bac3",
                                        "HG002_Revio.paraphase.json:md5,03ba9306a2bea70a817e4b2ef32d55b4",
                                        [
                                            "HG002_Revio_hba.vcf.gz:md5,fb0207f30c5bb0e3beba699e147feed9",
                                            "HG002_Revio_hba.vcf.gz.tbi:md5,63656171e041cf226e8b312c3e989eee"
                                        ]
                                    ]
                                ],
                                [
                                    "test_single.ped:md5,bd5cec27ba7337a85cf98e787131e2b5"
                                ],
                                [
                                    [
                                        [
                                            [
                                                "HG002_Revio_phased.vcf.gz:md5,cde30af0e4f160e91b9db0fcebebc2d1",
                                                "HG002_Revio_phased.vcf.gz.tbi:md5,3e2fc64bb2f893b44dcf24817b4632f3"
                                            ]
                                        ],
                                        [
                                            [
                                                "HG002_Revio_stats.blocks.tsv:md5,eaf8f60522034f52b3a45ced6ebed2ea",
                                                "HG002_Revio_stats.stats.tsv:md5,b571fcb3ef8a9e7921dd2824cfc1f549"
                                            ]
                                        ]
                                    ]
                                ],
                                [
                                    "nf_core_pipeline_software_mqc_versions.yml:md5,3675babbfa3ff94ad52a8d090e42545e",
                                    "params_2024-09-20_14-33-35.json:md5,ef85e454ae8828ddf71282d9aa407fb2"
                                ],
                                [
                                    [
                                        [
                                            [
                                                "HG002_Revio_cramino_aligned_phased.arrow:md5,72df2934ff8aa7e1bf8cf8a4881a0d2a",
                                                "HG002_Revio_cramino_aligned_phased.txt:md5,3af7900149aa0b16e1f858feed041aaa"
                                            ]
                                        ],
                                        [
                                            [
                                                "HG002_Revio_cramino_aligned.arrow:md5,72df2934ff8aa7e1bf8cf8a4881a0d2a",
                                                "HG002_Revio_cramino_aligned.txt:md5,10dc53594d12a53b793f8c8d4382b1fb"
                                            ]
                                        ]
                                    ],
                                    [
                                        [
                                            "HG002_Revio.mosdepth.global.dist.txt:md5,6186315d4d65eda85553af82a98829d1",
                                            "HG002_Revio.mosdepth.region.dist.txt:md5,c6c7ee8f056b8b2b92c97ec472b1db16",
                                            "HG002_Revio.mosdepth.summary.txt:md5,35c51f1ad9d2856d1e6446205f19a8e3",
                                            "HG002_Revio.regions.bed.gz:md5,e051f64c7a780d67ea6727a327dd4281",
                                            "HG002_Revio.regions.bed.gz.csi:md5,026eef1c69fb4aa3a1687463fe2088ab"
                                        ]
                                    ],
                                    [
                                        [
                                            [
                                                "test_single.html:md5,d05e0eceb70ada3a0c25f99a16ad1889",
                                                "test_single.pairs.tsv:md5,4a0988fc3c0fe5cfd5dd205fe6755595",
                                                "test_single.samples.tsv:md5,1685dc6cb8c6b9806ca636662980d686"
                                            ]
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "HG002_Revio_fastqc.html:md5,0337b9b2bb1368340c79624051d97363",
                                            "HG002_Revio_fastqc.zip:md5,0b4dde7cf60185c165e61b689f246787"
                                        ]
                                    ],
                                    [
                                        [
                                            "HG002_PacBio_Revio.fastq.gz.tsv.zst:md5,4b073293b3e771d19b4cfdb07909571b"
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            [
                                                "test_single_sniffles.vcf.gz:md5,8386b2d30b0e77a644f26c820013df3f",
                                                "test_single_sniffles.vcf.gz.tbi:md5,6fd9ceb01d7baec1b96ecb3b29583e0d"
                                            ]
                                        ],
                                        [
                                            [
                                                "HG002_Revio_sniffles.snf:md5,b2d5a63bb00cabcb2527d361721f6fa1",
                                                "HG002_Revio_sniffles.vcf.gz:md5,9b69d2d3f79130b6f483d12855ab07fa",
                                                "HG002_Revio_sniffles.vcf.gz.tbi:md5,4d5ed6747851d968a0b9b51d0e3558b1"
                                            ]
                                        ]
                                    ]
                                ]
                            ],
                            "single_end": false
                        },
                        "out.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,2ae45bee794ce6db65e7dc9c5cd036df"
                ],
                "csi": [
                    
                ],
                "tbi": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": [
                                [
                                    [
                                        "HG002_Revio_phased.bam:md5,64e1d1ed3d180589193c13f2c94d1098",
                                        "HG002_Revio_phased.bam.bai:md5,bb768f48be5144262c0b713ef183d152"
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "HG002_Revio.asm.bp.hap1.p_ctg.assembly_summary:md5,e4da3107f2421d42c6e2f59bff4e47d9",
                                            "HG002_Revio.asm.bp.hap1.p_ctg.fasta.gz:md5,eff16bf92abff02c36cceb2eb5a9871c",
                                            "HG002_Revio.asm.bp.hap2.p_ctg.assembly_summary:md5,bf908610472e602fd986b9d4a4d58b9e",
                                            "HG002_Revio.asm.bp.hap2.p_ctg.fasta.gz:md5,e28c886baffe24c1a2972a1d95b1b250"
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "HG002_Revio.dip.bed:md5,4e3cde029bceaa9363402e997f12f280",
                                            "HG002_Revio.dip.vcf.gz:md5,9259613d7b01347b2d817d4e547c4203",
                                            "HG002_Revio.hap1.bam:md5,5c739e8b3c2f659c3847498848280bbf",
                                            "HG002_Revio.hap1.bam.bai:md5,a57472fa9365470db09f5ea8c7a7420a",
                                            "HG002_Revio.hap1.bed:md5,3d75c9538e748a7672b934bdbde90df3",
                                            "HG002_Revio.hap1.paf.gz:md5,46a3056ecd833d9a932c53fc46a3cafc",
                                            "HG002_Revio.hap1.sam.gz:md5,0a161b44c29c0581ada87fee50be38c2",
                                            "HG002_Revio.hap1.var.gz:md5,173218f6c3f5f067e0932a8331f12e3d",
                                            "HG002_Revio.hap2.bam:md5,df63e7bb5dcee6ca5e731272dad0eb99",
                                            "HG002_Revio.hap2.bam.bai:md5,ed8184223edea0297c7f9283ea7927bf",
                                            "HG002_Revio.hap2.bed:md5,9ac411d49f08a8ec70e4792890e9faf9",
                                            "HG002_Revio.hap2.paf.gz:md5,a99e451e4af96a7ab120cedd49eea618",
                                            "HG002_Revio.hap2.sam.gz:md5,410acad899eda2f0a4a3e1eb4acf81cb",
                                            "HG002_Revio.hap2.var.gz:md5,53dc16fc893a529a95067da161b4dad6",
                                            "HG002_Revio.pair.vcf.gz:md5,a390558ab50ce0dc9efec070b1c70c22"
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "HG002_Revio.HG002_Revio.copynum.bedgraph:md5,517bc59c1b235490c79aa8319437b033",
                                            "HG002_Revio.HG002_Revio.depth.bw:md5,20757c985f4713b8568dee05765db701",
                                            "HG002_Revio.HG002_Revio.maf.bw:md5,4699d7a664277863f06eb48c3cba3c41",
                                            "HG002_Revio.HG002_Revio.vcf.gz:md5,771f7d9ecbd161dedfbdd053ceff0d52",
                                            "HG002_Revio.HG002_Revio.vcf.gz.tbi:md5,65d6930f79e626a5542091b7a419dc2b",
                                            "HG002_Revio.log:md5,3b47302d7e8e0ffac4841123de74ffb3"
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            [
                                                
                                            ],
                                            [
                                                
                                            ]
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        "HG002_Revio.paraphase.bam:md5,839cab6fb9b453b36975e9ab650d2552",
                                        "HG002_Revio.paraphase.bam.bai:md5,7146463306700bc589b4085c3404bac3",
                                        "HG002_Revio.paraphase.json:md5,03ba9306a2bea70a817e4b2ef32d55b4",
                                        [
                                            "HG002_Revio_hba.vcf.gz:md5,fb0207f30c5bb0e3beba699e147feed9",
                                            "HG002_Revio_hba.vcf.gz.tbi:md5,63656171e041cf226e8b312c3e989eee"
                                        ]
                                    ]
                                ],
                                [
                                    "test_single.ped:md5,bd5cec27ba7337a85cf98e787131e2b5"
                                ],
                                [
                                    [
                                        [
                                            [
                                                "HG002_Revio_phased.vcf.gz:md5,cde30af0e4f160e91b9db0fcebebc2d1",
                                                "HG002_Revio_phased.vcf.gz.tbi:md5,3e2fc64bb2f893b44dcf24817b4632f3"
                                            ]
                                        ],
                                        [
                                            [
                                                "HG002_Revio_stats.blocks.tsv:md5,eaf8f60522034f52b3a45ced6ebed2ea",
                                                "HG002_Revio_stats.stats.tsv:md5,b571fcb3ef8a9e7921dd2824cfc1f549"
                                            ]
                                        ]
                                    ]
                                ],
                                [
                                    "nf_core_pipeline_software_mqc_versions.yml:md5,3675babbfa3ff94ad52a8d090e42545e",
                                    "params_2024-09-20_14-33-35.json:md5,ef85e454ae8828ddf71282d9aa407fb2"
                                ],
                                [
                                    [
                                        [
                                            [
                                                "HG002_Revio_cramino_aligned_phased.arrow:md5,72df2934ff8aa7e1bf8cf8a4881a0d2a",
                                                "HG002_Revio_cramino_aligned_phased.txt:md5,3af7900149aa0b16e1f858feed041aaa"
                                            ]
                                        ],
                                        [
                                            [
                                                "HG002_Revio_cramino_aligned.arrow:md5,72df2934ff8aa7e1bf8cf8a4881a0d2a",
                                                "HG002_Revio_cramino_aligned.txt:md5,10dc53594d12a53b793f8c8d4382b1fb"
                                            ]
                                        ]
                                    ],
                                    [
                                        [
                                            "HG002_Revio.mosdepth.global.dist.txt:md5,6186315d4d65eda85553af82a98829d1",
                                            "HG002_Revio.mosdepth.region.dist.txt:md5,c6c7ee8f056b8b2b92c97ec472b1db16",
                                            "HG002_Revio.mosdepth.summary.txt:md5,35c51f1ad9d2856d1e6446205f19a8e3",
                                            "HG002_Revio.regions.bed.gz:md5,e051f64c7a780d67ea6727a327dd4281",
                                            "HG002_Revio.regions.bed.gz.csi:md5,026eef1c69fb4aa3a1687463fe2088ab"
                                        ]
                                    ],
                                    [
                                        [
                                            [
                                                "test_single.html:md5,d05e0eceb70ada3a0c25f99a16ad1889",
                                                "test_single.pairs.tsv:md5,4a0988fc3c0fe5cfd5dd205fe6755595",
                                                "test_single.samples.tsv:md5,1685dc6cb8c6b9806ca636662980d686"
                                            ]
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            "HG002_Revio_fastqc.html:md5,0337b9b2bb1368340c79624051d97363",
                                            "HG002_Revio_fastqc.zip:md5,0b4dde7cf60185c165e61b689f246787"
                                        ]
                                    ],
                                    [
                                        [
                                            "HG002_PacBio_Revio.fastq.gz.tsv.zst:md5,4b073293b3e771d19b4cfdb07909571b"
                                        ]
                                    ]
                                ],
                                [
                                    [
                                        [
                                            [
                                                "test_single_sniffles.vcf.gz:md5,8386b2d30b0e77a644f26c820013df3f",
                                                "test_single_sniffles.vcf.gz.tbi:md5,6fd9ceb01d7baec1b96ecb3b29583e0d"
                                            ]
                                        ],
                                        [
                                            [
                                                "HG002_Revio_sniffles.snf:md5,b2d5a63bb00cabcb2527d361721f6fa1",
                                                "HG002_Revio_sniffles.vcf.gz:md5,9b69d2d3f79130b6f483d12855ab07fa",
                                                "HG002_Revio_sniffles.vcf.gz.tbi:md5,4d5ed6747851d968a0b9b51d0e3558b1"
                                            ]
                                        ]
                                    ]
                                ]
                            ],
                            "single_end": false
                        },
                        "out.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2ae45bee794ce6db65e7dc9c5cd036df"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-24T16:40:24.837767435"
    }
}